staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1176,21 +1228,42 @@ type cacheWatcher struct {
 	// human readable identifier that helps assigning cacheWatcher
 	// instance with request
 	identifier string
+
+	// drainInputBuffer indicates whether we should delay closing this watcher
+	// and send all event in the input buffer.
+	drainInputBuffer bool
+
+	// bookmarkAfterResourceVersion holds an RV that indicates when we should start delivering bookmark events
```

Comment:
Brainstorming:

We effectively have a state machine here with states:
- WAITING_FOR_BOOKMARK
- BOOKMARK_RECEIVED
- BOOKMARK_SENT

I'm wondering if storing that wouldn't be more intuitive:
- we have a field "bookmarkAfterResourceVersion" (which is effectively immutable)
- we have a state (that is mutable):
  (a) in waiting state and got high-enough bookmark, we transition to boomark-received
  (b) if in bookmar-receivced state and sending bookmark transition to bookmar-sent state
  (c) bookmar-send is a terminal state

Reply:
Thanks for doing this!